### DISCUSSION TITLE
curl_ws_send / curl_ws_recv is a thread safety?

### DISCUSSION THREAD

**Discussion by devpreview** [curl/curl#10308] [Q&A] [answered]
If I call `curl_ws_send()` and `curl_ws_recv()` from different threads I have error:

[vtls/openssl.c:730](https://github.com/curl/curl/blob/master/lib/vtls/openssl.c#L730): bio_cf_in_read: Assertion `data' failed.

If I understand correctly, this is normal for OpenSSL with multithreading.

```
> curl --version  
curl 7.87.0-DEV (x86_64-pc-linux-gnu) libcurl/7.87.0-DEV OpenSSL/3.0.2 zlib/1.2.11 libpsl/0.21.0 (+libidn2/2.3.2) nghttp2/1.43.0
Release-Date: [unreleased]
Protocols: dict file ftp ftps gopher gophers http https imap imaps mqtt pop3 pop3s rtsp smb smbs smtp smtps telnet tftp ws wss
Features: alt-svc AsynchDNS Debug HSTS HTTP2 HTTPS-proxy IPv6 Largefile libz NTLM NTLM_WB PSL SSL threadsafe TLS-SRP TrackMemory UnixSockets
```

I have a simple hotfix for handle this issue:

```
> git diff lib/vtls/openssl.c
diff --git a/lib/vtls/openssl.c b/lib/vtls/openssl.c
index e7a1caabf..663e3fafb 100644
--- a/lib/vtls/openssl.c
+++ b/lib/vtls/openssl.c
@@ -727,7 +727,10 @@ static int bio_cf_in_read(BIO *bio, char *buf, int blen)
   ssize_t nread;
   CURLcode result = CURLE_RECV_ERROR;
 
-  DEBUGASSERT(data);
+  if(!data) {
+    BIO_set_retry_read(bio);
+    return 0;
+  }
   /* OpenSSL catches this case, so should we. */
   if(!buf)
     return 0;
```

**Comment by dfandrich** [comment, marked as answer]
You simply can't use the same handle from different threads. That hotfix won't solve
anything and will just hide future problems. See https://curl.se/libcurl/c/threadsafe.html

**Reply by devpreview** [reply]
Thanks!

**Reply by devpreview** [reply]
> You simply can't use the same handle from different threads.

Then how to send data with `curl_ws_send()`? This always need to stop `curl_ws_recv()`?

**Comment by devpreview** [comment]
I think I understand how it works.
My question is incorrect because curl_ws_send() / curl_ws_recv() cannot be thread-safe or not.
All depends on TLS-backend. For sample NSS already without anything required (but marked as depricated).

**Comment by dfandrich** [comment]
docs/WEBSOCKET.md describes two different ways of using the API, and neither of
them uses threads.
